elastic/kibana
# [Upgrade Assistant] Avoid creating the new index if it already exists when reindexing

Fix https://github.com/elastic/kibana/issues/123638

### Summary
If for any reason while we're reindexing and kibana server is stopped (somewhere between step 2 and 3) right after it created the new index but before it stored the current step in the SO, next time kibana starts it will resume from step 2 and will re-attempt to create the new index once again. This generates an error that wasn't being handled at the moment.

There's a [another issue](https://github.com/elastic/kibana/issues/123816) for making this flow more resilient to different kinds of failures.

**Note**: This assumes that the newly created index `reindexed-v7-*` was initially created by kibana for internal use only and not by the user, in which case could generate conflicts.

**How to test**

1. Start elasticsearch with `yarn es snapshot --license=trial -E path.data=./data-6.8.20` and kibana with `yarn start`. You can use this snapshot  I created [data-6.8.20.zip](https://github.com/elastic/kibana/files/7941871/data-6.8.20.zip) that contains a bunch of indices from 6.8.20 that require migration.
2. Navigate to `Stack Management` -> `Upgrade Assistant` -> `ES deprecations`
3. Start migrating one of the indices of the list and when it reaches step `3 - Reindex documents` kill kibana
4. Start kibana once again and go to console and run the following:
```
GET /.kibana/_search
{
  "query": {
    "term": {
      "type": {
        "value": "upgrade-assistant-reindex-operation"
      }
    }
  }
}
```
This will give you a list of the saved objects used by UA. Find the index that you started migrating and copy the documentID (which has the form `upgrade-assistant-reindex-operation:$HASH`). Run the following query to force the migration to be back at step 2 (make sure to set `lastCompletedStep to 20`)
```
PUT /.kibana/_doc/$DOCUMENT_ID
{
  "upgrade-assistant-reindex-operation" : {
      "indexName" : "kibana_sample_data_ecommerce",
      "newIndexName" : "reindexed-v7-kibana_sample_data_ecommerce",
      "status" : 0,
      "lastCompletedStep" : 20,
      "locked" : "2022-01-26T12:09:50+01:00",
      "reindexTaskId" : null,
      "reindexTaskPercComplete" : null,
      "errorMessage" : null,
      "runningReindexCount" : null
  },
  "type" : "upgrade-assistant-reindex-operation",
  "references" : [ ],
  "coreMigrationVersion" : "7.17.0",
  "updated_at" : "2022-01-26T11:09:50.234Z"
}
```
5. Re run the migration and verify that it all finishes successfully

Note: when this PR is merged, I'll create a new one for getting this change into main and 8.x
